homeassistant/components/balboa/climate.py

```diff
@@ -0,0 +1,208 @@
+"""Support for Balboa Spa Wifi adaptor."""
+from __future__ import annotations
+
+import math
+
+from homeassistant.components.climate import ClimateEntity
+from homeassistant.components.climate.const import (
+    CURRENT_HVAC_HEAT,
+    CURRENT_HVAC_IDLE,
+    FAN_HIGH,
+    FAN_LOW,
+    FAN_MEDIUM,
+    FAN_OFF,
+    HVAC_MODE_AUTO,
+    HVAC_MODE_HEAT,
+    HVAC_MODE_OFF,
+    SUPPORT_FAN_MODE,
+    SUPPORT_PRESET_MODE,
+    SUPPORT_TARGET_TEMPERATURE,
+)
+from homeassistant.const import (
+    ATTR_TEMPERATURE,
+    PRECISION_HALVES,
+    PRECISION_WHOLE,
+    TEMP_CELSIUS,
+    TEMP_FAHRENHEIT,
+)
+
+from .balboa_entity import BalboaEntity
+from .const import CLIMATE, CLIMATE_SUPPORTED_FANSTATES, CLIMATE_SUPPORTED_MODES
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the spa climate device."""
+    async_add_entities([BalboaSpaClimate(hass, entry, CLIMATE)], True)
+
+
+class BalboaSpaClimate(BalboaEntity, ClimateEntity):
+    """Representation of a Balboa Spa Climate device."""
+
+    _attr_icon = "mdi:hot-tub"
+    _attr_fan_modes = CLIMATE_SUPPORTED_FANSTATES
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        features = SUPPORT_TARGET_TEMPERATURE | SUPPORT_PRESET_MODE
+
+        if self._client.have_blower():
+            features |= SUPPORT_FAN_MODE
+
+        return features
+
+    @property
+    def hvac_modes(self) -> list[str]:
+        """Return the list of supported HVAC modes."""
+        if self._client.get_heatmode() == self._client.HEATMODE_RNR:
+            return [*CLIMATE_SUPPORTED_MODES, HVAC_MODE_AUTO]
+        return CLIMATE_SUPPORTED_MODES
+
+    @property
+    def hvac_mode(self) -> str:
+        """Return the current HVAC mode."""
+        mode = self._client.get_heatmode()
+        if mode == self._client.HEATMODE_READY:
+            return HVAC_MODE_HEAT
+        if mode == self._client.HEATMODE_RNR:
+            return HVAC_MODE_AUTO
+        return HVAC_MODE_OFF
```

Comment:
Use a `dict` lookup instead

Reply:
I think this one is problematic.  the heatmodes are just integers. 0, 1, 2.  While I could make an array, I feel like that is an issue where you are trusting that the internal constants of the backend library don't change on you.  (yes, I know I'm the author of the backend library here, but still)  I'm expressly exporting them in the library so you can use those and not have to care what the internal values are.

Reply:
For this one, and the blower map, I feel like no matter how I convert this into a dict, I'm relying on the order of internal library constants to never change.  For both of these, the values are integers, so you could map them to an array inside HA, but then if the backend library ends up having to change one of the values for some reason, then the array doesn't match up any more.

I could switch to using the text mode of the library and return "Ready", but again, then I'm just mapping words to that internal integer array in the library, and hoping it never changes.

As written, if the library changes any of these constants around, everything keeps working.  This feels like a cosmetic change that will actually make maintenance harder in the future potentially, because I'm just hoping the library never changes underneath me.

Reply:
Additionally, I could pull the string list from the library, but then I'd have a big ugly chunk of code in the init that maps words back to integers just so I could make this section slightly shorter.

Reply:
Use the values to build the dict. Event if the constant values change it will keep working.

Reply:
In https://github.com/garbled1/pybalboa/blob/master/pybalboa/balboa.py#L95 you could move the constants above the constructor and then do 

BALBOA_HVAC_MODE_TO_HASS = {
    BalboaSpaWifi.HEATMODE_READY: HVAC_MODE_HEAT,
    BalboaSpaWifi.HEATMODE_RNR: HVAC_MODE_AUTO
}

Then

```
    return BALBOA_HVAC_MODE_TO_HASS.get(mode, HVAC_MODE_OFF)
```

Reply:
Another example here https://github.com/home-assistant/core/blob/dev/homeassistant/components/nexia/climate.py#L94

All of these avoid linear searches

Reply:
OK yeah.. I'll build this without changing the library just yet, and then modify in a future PR.  I don't want to break the HACS version or try to dual-maintain them until I have the whole integration (with all the other sensors/binary/etc) loaded in to core.  If I change the library like that, then I'll have to change both this and the HACS version and its just pain.